client/src/main/java/io/pravega/client/connection/impl/CommandEncoder.java

```diff
@@ -0,0 +1,419 @@
+            }
+        }
+
+        /**
+         * Write/flush given data to network channel by writing the Append block End containing the given/input data
+         * and indicating the size of previous append block data.
+         *
+         * @param sizeOfWholeEvents Size of data followed by this append block end.
+         * @param data              Remaining data
+         */
+        private void flushToBuffer(int sizeOfWholeEvents, ByteBuf data) {
+            writeMessage(new AppendBlockEnd(id, sizeOfWholeEvents, data, eventCount, lastEventNumber, requestId), buffer);
+            eventCount = 0;
+        }
+    }
+    
+    /**
+     * Write/flush Multi session's buffered data to network channel buffer by iterating over all the pending sessions .
+     */
+    private void flushAllToBuffer() {
+        if (!pendingWrites.isEmpty()) {
+            ArrayList<Session> sessions = new ArrayList<>(pendingWrites.values());
+            sessions.forEach(session -> session.flushToBuffer());
+        }
+    }
+    
+    @Synchronized
+    public void write(WireCommand msg) throws IOException {
+        if (msg instanceof SetupAppend) {
+            breakCurrentAppend();
+            flushAllToBuffer();
+            writeMessage((SetupAppend) msg, buffer);
+            SetupAppend setup = (SetupAppend) msg;
+            setupSegments.put(new SimpleImmutableEntry<>(setup.getSegment(), setup.getWriterId()),
+                              new Session(setup.getWriterId(), setup.getRequestId()));
+            flushBuffer();
+        } else if (msg instanceof Hello) {
+            Preconditions.checkState(isChannelFree());
+            Preconditions.checkState(pendingWrites.isEmpty());
+            writeMessage((WireCommand) msg, buffer);
+            flushBuffer();
+        } else {
+            breakCurrentAppend();
+            flushAllToBuffer();
+            writeMessage((WireCommand) msg, buffer);
+            flushBuffer();
+        }
+    }
+    
+    @Synchronized
+    public void write(Append append) throws IOException {
+        Session session = setupSegments.get(new SimpleImmutableEntry<>(append.getSegment(), append.getWriterId()));
+        validateAppend(append, session);
+        final ByteBuf data = append.getData().slice();
+        final AppendBatchSizeTracker blockSizeSupplier = (appendTracker == null) ? null :
+                appendTracker.apply(append.getFlowId());
+
+        if (blockSizeSupplier != null) {
+            blockSizeSupplier.recordAppend(append.getEventNumber(), data.readableBytes());
+        }
+        if (isChannelFree()) {
+            if (session.isFree()) {
+                session.record(append);
+                startAppend(blockSizeSupplier, append);
+                continueAppend(data);
+                if (bytesLeftInBlock == 0) {
+                    completeAppend(null);
+                    flushBuffer();
+                }
+            } else {
+                session.record(append);
```

Comment:
I'm not sure this line is actually reachable. 
See discussion here: https://github.com/pravega/pravega/pull/4121#discussion_r330332792
from when it was added.